### PR TITLE
Change hier singleton to hierarchy

### PR DESCRIPTION
**Pull request by entro-pi**
This change was decided upon as the hierarchy naming convention
is less confusing than the current hier convention.

I'm a bit unfamiliar with unit testing in an environment such as this.
As well, kubernetes in general is new to me. @tashimi how would one
go about testing my changes?

This is my very first PR, please let me know if I've tripped up anywhere. (as one would anyway)

### CODE REVIEW

**Comment by entro-pi** [comment]
@tashimi Nevermind! The robot got me the information I need to continue testing.

**Comment by adrianludwin** [comment]
Can you join the Kubernetes Slack so we can discuss? Join at slack.k8s.io,
my user name there is just "Adrian Ludwin".

File: incubator/hnc/go.mod [hunk 1]
```diff
@@ -14,4 +14,5 @@ require (
 	k8s.io/apimachinery v0.0.0-20190404173353-6a84e37a896d
 	k8s.io/client-go v11.0.1-0.20190409021438-1a26190bd76a+incompatible
 	sigs.k8s.io/controller-runtime v0.2.2
+	sigs.k8s.io/controller-tools v0.2.0-beta.4 // indirect
```
Comment (adrianludwin) [inline incubator/hnc/go.mod]: Can you remove this? Read about Go modules to see why this was added, but basically you must have called `go get controller-tools` while in this directory.

File: incubator/hnc/go.sum [hunk 2]
```diff
@@ -33,20 +33,27 @@ github.com/dgrijalva/jwt-go v3.2.0+incompatible/go.mod h1:E3ru+11k8xSBh+hMPgOLZm
 github.com/docopt/docopt-go v0.0.0-20180111231733-ee0de3bc6815/go.mod h1:WwZ+bS3ebgob9U8Nd0kOddGdZWjyMGR8Wziv+TBNwSE=
 github.com/evanphx/json-patch v4.5.0+incompatible h1:ouOWdg56aJriqS0huScTkVXPC5IcNrDCXZ6OoTAWu7M=
 github.com/evanphx/json-patch v4.5.0+incompatible/go.mod h1:50XU6AFN0ol/bzJsmQLiYLvXMP4fmwYFNcr97nuDLSk=
+github.com/fatih/color v1.7.0 h1:DkWD4oS2D8LGGgTQ6IvwJJXSL5Vp2ffcQg58nFV38Ys=
```
Comment (adrianludwin) [inline incubator/hnc/go.sum]: You should be able to restore this file too

File: incubator/hnc/pkg/controllers/hierarchy_controller_test.go [hunk 3]
```diff
@@ -27,23 +27,23 @@ var _ = Describe("Hierarchy", func() {
 	})
 
 	It("should set a child on the parent", func() {
-		fooHier := newHierarchy(fooName)
-		fooHier.Spec.Parent = barName
-		updateHierarchy(ctx, fooHier)
+		fooHierarchy := newHierarchy(fooName)
```
Comment (adrianludwin) [inline incubator/hnc/pkg/controllers/hierarchy_controller_test.go]: Please don't change the names of the variables for a few reasons:

1. It makes it much harder than it needs to be to review the relevant parts of this change.
2. It will make it more difficult to track changes that happened before and after this change.
3. Golang code style is to use short local variable names.

**Comment by entro-pi** [comment]
Righty-o! I was a little unsure that I was doing the correct task in the
first place. Thank you for the clarification!

I will complete this task as soon as possible.

Cheers!

File: incubator/hnc/go.mod (hunk 1 above)
Comment (entro-pi) [reply incubator/hnc/go.mod]: Thanks! I was unsure as to exactly what this was. I'll do that reading in a bit. Go mods are something I haven't used just yet in my Go career.

File: incubator/hnc/go.sum (hunk 2 above)
Comment (entro-pi) [reply incubator/hnc/go.sum]: Finished this as well.

File: incubator/hnc/pkg/controllers/hierarchy_controller_test.go (hunk 3 above)
Comment (entro-pi) [reply incubator/hnc/pkg/controllers/hierarchy_controller_test.go]: Here we go. let me know if I missed anything.

**Comment by entro-pi** [comment]
Made the changes requested, as well as reverting the changes made inadvertently. 
9/10 tests pass, if i could get some help with figuring out how to get it to pass that'd be awesome.

Summarizing 1 Failure:

[Fail] Secret [It] should not be propagated if marked as modified 
~/Code/go/src/github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/pkg/controllers/object_controller_test.go:90

**Comment by entro-pi** [comment]
Nevermind! Got it to pass.

**Comment by adrianludwin** [comment]
Whoops sorry I didn't realize this before, but you're not supposed to hand-edit the YAML files. Instead, in the `hierarchy_types.go` file, you need to change the Go structs to the correct type (eg from `Hierarchy` to `HierarchyConfig`; from `HierarchySpec` to `HierarchyConfigSpec`, etc). This will mean you need to change where they're referenced in the rest of the HNC as well.

Then, run `make manifests` to _automatically_ regenerate all the YAML files.

Finally, can you please squash all your commits into one? `git rebase -i master` ought to do it. Then you just force-push to the same branch in your fork, and it will show up here.

**Comment by adrianludwin** [comment]
Also, you'll probably want to pull from the latest `master` in the upstream repo as there have been a bunch of changes to both `api/.../hierarchy_types.go` and `pkg/controllers/hierarchy_controller.go`.

**Comment by entro-pi** [comment]
Cool! Thank you! I'll work on that this afternoon.

File: incubator/hnc/go.mod (hunk 1 above)
Comment (adrianludwin) [inline incubator/hnc/go.mod]: Please revert the changes to `go.mod` and `go.sum`, these just happened because you called `go get controller-tools` while in this directory but we don't actually want a dependency on them.

File: incubator/hnc/pkg/controllers/hierarchy_controller.go [hunk 4]
```diff
@@ -41,7 +41,7 @@ import (
 // as well as ensuring all objects in the forest are propagated correctly when the hierarchy
 // changes. It can also set the status of the Hierarchy CRs, as well as (in rare cases) override
 // part of its spec (i.e., if a parent namespace no longer exists).
-type HierarchyReconciler struct {
+type HierarchyConfigReconciler struct {
```
Comment (adrianludwin) [inline incubator/hnc/pkg/controllers/hierarchy_controller.go]: As mentioned above, please revert this change - it's only the API names that are changing, this is just an internal data structure.

File: incubator/hnc/pkg/kubectl/tree.go [hunk 5]
```diff
@@ -52,7 +52,7 @@ var treeCmd = &cobra.Command{
 	},
 }
 
-func printSubtree(prefix string, hier *tenancy.Hierarchy) {
+func printSubtree(prefix string, hier *tenancy.HierarchyConfig) {
```
Comment (adrianludwin) [inline incubator/hnc/pkg/kubectl/tree.go]: Did this compile?

File: incubator/hnc/pkg/validators/hierarchy.go [hunk 6]
```diff
@@ -27,22 +27,22 @@ const (
 //
 // +kubebuilder:webhook:path=/validate-hnc-x-k8s-io-v1alpha1-hierarchies,mutating=false,failurePolicy=fail,groups="hnc.x-k8s.io",resources=hierarchies,verbs=create;update,versions=v1alpha1,name=hierarchies.hnc.x-k8s.io
 
-type Hierarchy struct {
+type HierarchyConfig struct {
```
Comment (adrianludwin) [inline incubator/hnc/pkg/validators/hierarchy.go]: Revert (as mentioned)

File: incubator/hnc/scripts/hier-set-parent [hunk 7]
```diff
@@ -1,9 +1,9 @@
 #!/bin/bash
 cat <<EOF | kubectl apply -f -
 apiVersion: hnc.x-k8s.io/v1alpha1
-kind: Hierarchy
+kind: HierarchyConfig
```
Comment (adrianludwin) [inline incubator/hnc/scripts/hier-set-parent]: Remember that this should be `HierarchyConfiguration`

File: incubator/hnc/go.mod (hunk 1 above)
Comment (entro-pi) [reply incubator/hnc/go.mod]: Oh, my, glob. That took some git magic. I learned a lot! Thanks!

File: incubator/hnc/api/v1alpha1/hierarchy_types.go [hunk 8]
```diff
@@ -38,7 +38,7 @@ type Hierarchy struct {
 }
 
 // HierarchySpec defines the desired state of Hierarchy
-type HierarchySpec struct {
+type HierarchyConfigSpec struct {
```
[the remark this answers is not in the document]
Comment (entro-pi) [reply incubator/hnc/api/v1alpha1/hierarchy_types.go]: Ok! Got this.

I think what was causing my problems that cause it not to compile have been popping up after I fix merge conflicts from upstream on the push.

Lesson learned, make sure you pull from upstream (NOT fork) before pushing.
Resolved.

File: incubator/hnc/pkg/controllers/hierarchy_controller.go (hunk 4 above)
Comment (entro-pi) [reply incubator/hnc/pkg/controllers/hierarchy_controller.go]: Another thing I've learned is the power of the namespace. When I started this I was under the assumption that if they were named close to the same they were part of the same process. This shows it not to be so, and teaches me more about how unit testing works. Ie mixing types and seeing what blows up.

File: incubator/hnc/pkg/kubectl/tree.go (hunk 5 above)
Comment (entro-pi) [reply incubator/hnc/pkg/kubectl/tree.go]: Nope!

It does now.
Resolved.

File: incubator/hnc/pkg/validators/hierarchy.go (hunk 6 above)
Comment (entro-pi) [reply incubator/hnc/pkg/validators/hierarchy.go]: Done.

File: incubator/hnc/scripts/hier-set-parent (hunk 7 above)
Comment (entro-pi) [reply incubator/hnc/scripts/hier-set-parent]: So noted, and changed.

Resolved.

File: incubator/hnc/cmd/manager/main.go [hunk 9]
```diff
@@ -84,7 +84,7 @@ func main() {
 	// Create validating admission controllers
 	if !novalidation {
 		setupLog.Info("Registering validating webhook (won't work when running locally; use --novalidation)")
-		mgr.GetWebhookServer().Register(validators.HierarchyServingPath, &webhook.Admission{Handler: &validators.Hierarchy{
+		mgr.GetWebhookServer().Register(validators.HierarchyServingPath, &webhook.Admission{Handler: &validators.HierarchyConfig{
```
[the remark this answers is not in the document]
Comment (entro-pi) [reply incubator/hnc/cmd/manager/main.go]: Again, I learned a lot over these few days. I understand the code enough to know why this is wrong, and have changed it back.

Cheers!

**Comment by entro-pi** [comment]
I've managed to squash the commits into one big commit and one merge
commit. I ended up making a new branch and merged it with my working one.
I'm unsure how to squash the merge commit? I tried rebase -i HEAD~1 and a
few other things but can't get it just right.

As for the Test line do I just put the scripts I've done? Ie

Test: test manifests hier-show

As for hier-show it'll take me a little while longer to properly read up on
kubectl to find out if it is running correctly or not.

**Comment by entro-pi** [comment]
hold up, pushed prematurely

**Comment by entro-pi** [comment]
> hold up, pushed prematurely

Nevermind, forgot to run make manifest before make test. It passes now.

**Comment by adrianludwin** [comment]
This looks almost ready. Last steps:

1. Can you please edit the commit message down to something shorter? You don't have to have the _history_ of your work on this PR, just tell people what they need to know about it in its current form.

2. Can you please test at least one of the kubectl commands as well? Use `source devenv` to set up your `PATH` env var, `make kubectl` to build the `/bin/kubectl-hnc` binary, and then `kubectl hnc tree <namespace>` to display a tree. Once you've done this, please add the fact that you tested this to the commit message and re-push.

Thanks!

**Comment by adrianludwin** [comment]
Looks good! For the record, this PR addresses #88. Since this has already been discussed in the MTWG, I'm going to go ahead and approve this myself rather than wait for further reviews.

/approve
/lgtm
